using accumulate type to do the computation of mean reduce(CPU)

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97351

 This PR will use accumulate type to do the computation of mean reduce for CPU path as GPU path.

cc @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

XiaobingSuper committed:
using accumulate type to do the computation of mean reduce(CPU)

[ghstack-poisoned]

peterbell10 commented on aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:
```diff
@@ -136,15 +136,34 @@ static void logcumsumexp_cpu_kernel(Tensor& result, const Tensor& self, int64_t
-    scalar_t factor = scalar_t(iter.num_output_elements()) / scalar_t(iter.numel());
-    binary_kernel_reduce(
+template <typename scalar_t, typename acc_t=scalar_t, typename out_t=scalar_t>
+void mean_kernel_impl(TensorIterator& iter) {
```
FYI `MeanOps` isn't actually used on CPU, see `TORCH_IMPL_FUNC(mean_out)`
https://github.com/pytorch/pytorch/blob/364d92f9b6864ce284fa13519c7ca5c87460e477/aten/src/ATen/native/ReduceOps.cpp#L1318

jgong5 replied:
I guess we'd have to make `binary_kernel_reduce_vec` work with `mean` first. Then, we don't have to invoke `sum_out` + `div_` there.

XiaobingSuper replied:
Currently, it may be better to call ```mean_stub``` path for bfloat16 and half, I met an accuracy issue when comparing ```nn.AdaptiveAvgPool2d```(calling mean) with inductor(using fused kernel, sum+div):
```
import torch
import torch._dynamo

class Model(torch.nn.Module):
    def __init__(self):
        super(Model, self).__init__()
        self.avgpool = torch.nn.AdaptiveAvgPool2d((1, 1))

    def forward(self, x):
        x = self.avgpool(x)
        return x

model = Model().eval()
opt_model = torch._dynamo.optimize('inductor')(model)

batch_size = 4
x = torch.randn(batch_size, 512, 7, 7).to(dtype=torch.bfloat16)
ref = model(x)
with torch.no_grad():
    for i in range(3):
        out = opt_model(x)

print((ref - out).max())
tol = 0.001
print(ref-out)
print(torch.allclose(ref, out, atol=tol, rtol=tol))
````

before this PR, this case always failed, the eager model seems to have a big difference with inductor result, @mingfeima has the plan to support the reduction vectorization kernel.

peterbell10 commented on aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:
```diff
@@ -136,15 +136,34 @@ static void logcumsumexp_cpu_kernel(Tensor& result, const Tensor& self, int64_t
+    mean_kernel_impl<at::Half, float>(iter);
+  } else if (iter.dtype(1) == kHalf && iter.dtype() == kFloat) {
+    // type promotion that does cast and reduction in a single kernel
+    mean_kernel_impl<at::Half, float, float>(iter);
```
This special case only exists for CUDA, see `make_reduction_from_out_ty`:
https://github.com/pytorch/pytorch/blob/364d92f9b6864ce284fa13519c7ca5c87460e477/aten/src/ATen/native/ReduceOpsUtils.h#L434-L442

XiaobingSuper replied:
removed.

XiaobingSuper committed:
Update on "using accumulate type to do the computation of mean reduce(CPU)"


 This PR will use accumulate type to do the computation of mean reduce for CPU path as GPU path.

cc jgong5 mingfeima sanchitintel ashokei jingxu10

[ghstack-poisoned]

XiaobingSuper committed:
Update on "using accumulate type to do the computation of mean reduce(CPU)"


 This PR will use accumulate type to do the computation of mean reduce for CPU path as GPU path.

cc jgong5 mingfeima sanchitintel ashokei jingxu10

[ghstack-poisoned]

jgong5 commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -1304,7 +1304,7 @@ TORCH_IMPL_FUNC(mean_out)
   // for production workloads. Once that's fixed, the following code can be used
   // in lieu of the sum + divide implementation below.
-  if (self.device().is_cpu()) {
+  if (self.device().is_cpu() && dtype != kHalf && dtype != kBFloat16) {
```
How about adding a note here explaining why you exclude bfloat16 and half here?

XiaobingSuper replied:
Yes, added.

XiaobingSuper commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -1304,7 +1304,9 @@ TORCH_IMPL_FUNC(mean_out)
-  if (self.device().is_cpu()) {
+  // Note: there has an accuracy loss for half and bfloat16 which has one more type
+  // cast compared with mean_stub path.
+  if (self.device().is_cpu() && dtype != kHalf && dtype != kBFloat16) {
```
For ComplexHalf and ComplexBfloat16, both cuda and CPU seems doesn't use accumulate dtype for mean_stub, it may be better to change it for higher precision.

XiaobingSuper committed:
Update on "using accumulate type to do the computation of mean reduce(CPU)"


 This PR will use accumulate type to do the computation of mean reduce for CPU path as GPU path.

cc jgong5 mingfeima sanchitintel ashokei jingxu10

[ghstack-poisoned]

XiaobingSuper committed:
Update on "using accumulate type to do the computation of mean reduce(CPU)"


 This PR will use accumulate type to do the computation of mean reduce for CPU path as GPU path.

cc jgong5 mingfeima sanchitintel ashokei jingxu10

[ghstack-poisoned]

XiaobingSuper commented on aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:
```diff
@@ -136,15 +136,28 @@ static void logcumsumexp_cpu_kernel(Tensor& result, const Tensor& self, int64_t
 
-static void mean_kernel_impl(TensorIterator& iter) {
-  AT_DISPATCH_ALL_TYPES_AND_COMPLEX(iter.dtype(), "mean_cpu", [&] {
-    scalar_t factor = scalar_t(iter.num_output_elements()) / scalar_t(iter.numel());
```
Cleared.

XiaobingSuper committed:
Update on "using accumulate type to do the computation of mean reduce(CPU)"


 This PR will use accumulate type to do the computation of mean reduce for CPU path as GPU path.

cc jgong5 mingfeima sanchitintel ashokei jingxu10

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot merge

mingfeima commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -1304,7 +1304,9 @@ TORCH_IMPL_FUNC(mean_out)
-  if (self.device().is_cpu()) {
+  // Note: there has an accuracy loss for half and bfloat16 which has one more type
+  // cast compared with mean_stub path.
+  if (self.device().is_cpu() && dtype != kHalf && dtype != kBFloat16) {
```
This is going to drop `BFloat16` performance in mean, right ?
So this patch is to fix the numerical issue ?

XiaobingSuper replied:
Yes, this pathc is to fix the numerical issue.

mingfeima commented:
@XiaobingSuper this is a workaround for now, consider add a test case to track this issue.

this one should be related to https://github.com/pytorch/pytorch/issues/91597
i will fix both issues later on.

XiaobingSuper committed:
Update on "using accumulate type to do the computation of mean reduce(CPU)"


 This PR will use accumulate type to do the computation of mean reduce for CPU path as GPU path.

cc jgong5 mingfeima sanchitintel ashokei jingxu10

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge
